src/Form/BasicSetup.php

```diff
@@ -0,0 +1,524 @@
+      }
+
+      // Check password. (from user.module user_validate_current_pass()).
+      //      require_once DRUPAL_ROOT . '/' . variable_get('password_inc', 'includes/password.inc');
+      //      $current_pass = user_check_password($values['current_pass'], $account);
+      //      if (!$current_pass) {
+      //        form_set_error('current_pass', t("Incorrect password."));
+      //      }.
+      return;
+    }
+    // Handle first step of SMS setup.
+    elseif (isset($values['sms_number'])) {
+      // Validate number.
+      $number = $values['sms_number'];
+      $number_errors = tfa_basic_valid_number($number);
+      if (!empty($number_errors)) {
+        foreach ($number_errors as $error) {
+          $form_state->setErrorByName('number', $error);
+        }
+      }
+      return;
+    }
+    // Validate plugin form.
+    elseif (!empty($storage['step_method'])) {
+      $method = $storage['step_method'];
+      $tfa_setup = $storage[$method];
+      if (!$tfa_setup->validateForm($form, $form_state)) {
+        foreach ($tfa_setup->getErrorMessages() as $element => $message) {
+          $form_state->setErrorByName($element, $message);
+        }
+      }
+    }
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function cancelForm(array &$form, FormStateInterface $form_state) {
+    $account = $form['account']['#value'];
+    drupal_set_message('TFA setup canceled.', 'warning');
+    $form_state->setRedirect('tfa.overview', ['user' => $account->id()]);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function submitForm(array &$form, FormStateInterface $form_state) {
+    $account = $form['account']['#value'];
+    $storage = $form_state->getStorage();
+    $values = $form_state->getValues();
+
+    // Password validation.
+    if (isset($values['current_pass'])) {
+      $storage['pass_confirmed'] = TRUE;
+      $form_state->setRebuild();
+      $form_state->setStorage($storage);
+      return;
+    }
+    // Submitting mobile number step.
+    elseif (!empty($values['sms_number'])) {
+      // Send code to number.
+      $storage['sms_number'] = $values['sms_number'];
+      $context = array('uid' => $account->id(), 'mobile_number' => $storage['sms_number']);
+      $setup_plugin = new TfaBasicSmsSetup($context, $storage['sms_number']);
+      $tfa_setup = new TfaSetup($setup_plugin, $context);
+      $tfa_setup->begin();
+      $errors = $tfa_setup->getErrorMessages();
+      if (!empty($errors)) {
+        foreach ($errors as $error) {
+          $form_state->setErrorByName('number', $error);
+        }
+      }
+      else {
+        // No errors so store setup.
+        $storage['tfa_basic_sms'] = $tfa_setup;
+      }
+      $form_state->setRebuild();
+      $form_state->setStorage($storage);
+      return;
+    }
+    // Disabling SMS delivery.
+    if (isset($values['sms_disable']) && $values['op'] === $values['sms_disable']) {
+      tfa_setup_save_data($account, array('sms' => FALSE));
```

Comment:
Another weird procedural callback. If this is a reusable function it goes into a service, otherwise a function of this class or a function in an entity